controller/src/main/java/io/pravega/controller/store/stream/ZKStream.java

```diff
@@ -605,14 +644,41 @@
 
     @Override
     public void refresh() {
-        // refresh all mutable records
-        cache.invalidateCache(statePath);
-        cache.invalidateCache(configurationPath);
-        cache.invalidateCache(truncationPath);
-        cache.invalidateCache(epochRecordPathFormat);
-        cache.invalidateCache(committingTxnsPath);
-        cache.invalidateCache(currentEpochRecordPath);
-        cache.invalidateCache(currentEpochRecordPath);
+        String id = this.idRef.getAndSet(null);
+        id = id == null ? "" : id;
+        // invalidate all mutable records in the cache 
+        store.invalidateCache(statePath, id);
+        store.invalidateCache(configurationPath, id);
+        store.invalidateCache(truncationPath, id);
+        store.invalidateCache(epochTransitionPath, id);
+        store.invalidateCache(committingTxnsPath, id);
+        store.invalidateCache(currentEpochRecordPath, id);
+    }
+
+    /**
+     * Method to retrieve unique Id for the stream. We use streamPosition as the unique id for the stream.
+     * If the id had been previously retrieved then this method simply returns
+     * the previous value else it retrieves the stored stream position from zookeeper. 
+     * The id of a stream is fixed for lifecycle of a stream and only changes when the stream is deleted and recreated. 
+     * The id is used for caching entities and safeguarding against stream recreation. 
+     * @return CompletableFuture which when completed contains stream's position as the id. 
+     */
+    private CompletableFuture<String> getId() {
+        String id = this.idRef.get();
+        if (!Strings.isNullOrEmpty(id)) {
+            return CompletableFuture.completedFuture(id);
+        } else {
+            // We will return empty string as id if the position does not exist. 
```

Comment:
its just going to result in value being loaded into cache against an empty id + path. which is harmless. when id becomes available, it will get cached against key with id and previous cache key will eventually be evicted from the cache.